File: packages/integrity/android/src/main/java/io/capkit/integrity/IntegrityPlugin.kt
package io.capkit.integrity

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.Uri
import android.os.Build
import androidx.core.content.ContextCompat
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import io.capkit.integrity.config.Config
import io.capkit.integrity.error.ErrorMessages
import io.capkit.integrity.error.NativeError
import io.capkit.integrity.logger.Logger
import io.capkit.integrity.models.CheckOptions
import io.capkit.integrity.utils.Utils

/**
 * Capacitor bridge for the Integrity plugin (Android).
 *
 * CONTRACT:
 * - This class is the ONLY entry point from JavaScript.
 * - All PluginCall instances MUST be resolved or rejected exactly once.
 *
 * Responsibilities:
 * - Parse JavaScript input
 * - Invoke the native implementation
 * - Resolve or reject PluginCall exactly once
 * - Map native NativeError to JS-facing error codes
 *
 * Forbidden:
 * - Platform-specific business logic
 * - Direct system API usage outside lifecycle-bound orchestration
 * - Throwing uncaught exceptions
 */
@CapacitorPlugin(
  name = "Integrity",
  permissions = [
    Permission(
      alias = "network",
      strings = [android.Manifest.permission.INTERNET],
    ),
  ],
)
class IntegrityPlugin : Plugin() {
  // ---------------------------------------------------------------------------
  // Properties
  // ---------------------------------------------------------------------------

  /**
   * Immutable plugin configuration.
   *
   * CONTRACT:
   * - Initialized exactly once in `load()`
   * - Treated as read-only afterward
   * - MUST NOT be mutated at runtime
   * - MUST NOT be accessed by the Impl layer
   */
  private lateinit var config: Config

  /**
   * Native implementation layer.
   *
   * CONTRACT:
   * - Owned by the Plugin layer
   * - Lifetime == plugin lifetime
   * - MUST NOT access PluginCall or Capacitor APIs
   * - MUST NOT perform UI operations
   */
  private lateinit var implementation: Integrity

  // ---------------------------------------------------------------------------
  // Event-related properties
  // ---------------------------------------------------------------------------

  /**
   * In-memory buffer for integrity signals detected before
   * a JavaScript listener is registered.
   *
   * Thread-safe implementation to prevent race conditions during
   * asynchronous event emission.
   */
  private val bufferedSignals = java.util.Collections.synchronizedList(mutableListOf<JSObject>())

  /**
   * BroadcastReceiver used to observe passive system events
   * relevant for integrity monitoring.
   */
  private lateinit var eventReceiver: IntegrityEventReceiver

  private companion object {
    /**
     * Canonical event name emitted to the JavaScript layer.
     *
     * CONTRACT:
     * - MUST remain stable across releases
     * - MUST match the JS-side event subscription name
     */
    private const val EVENT_INTEGRITY_SIGNAL = "integritySignal"
  }

  // ---------------------------------------------------------------------------
  // Lifecycle
  // ---------------------------------------------------------------------------

  /**
   * Called once when the plugin is loaded by the Capacitor bridge.
   *
   * CONTRACT:
   * - Called exactly once
   * - This is the ONLY valid place to:
   *   - read static configuration
   *   - initialize the native implementation
   *   - inject configuration into the implementation
   *   - register system event listeners (BroadcastReceivers)
   *
   * WARNING:
   * - Re-initializing config or implementation outside this method
   *   is considered a plugin defect.
   */
  override fun load() {
    super.load()

    config = Config(this)
    implementation = Integrity(context)
    implementation.updateConfig(config)

    Logger.debug("Plugin loaded. Version: ", BuildConfig.PLUGIN_VERSION)

    registerEventReceiver()
  }

  /**
   * Custom addListener implementation to flush early boot signals.
   *
   * PURPOSE:
   * - Ensures that signals captured during the early boot phase or
   * while the app was in the background are delivered as soon as
   * the JavaScript side registers a listener.
   */
  @PluginMethod
  override fun addListener(call: PluginCall) {
    val eventName = call.getString("eventName")
    super.addListener(call)

    if (eventName == EVENT_INTEGRITY_SIGNAL) {
      flushBufferedSignals()
    }
  }

  /**
   * Called when the host Activity resumes.
   *
   * PURPOSE:
   * - Flush any integrity signals captured while no JS listeners
   * were registered.
   * - Perform a targeted check for debugger attachment during resume.
   */
  override fun handleOnResume() {
    super.handleOnResume()
    flushBufferedSignals()

    // Immediate check for debugger attachment upon resume (Real-time monitor)
    if (android.os.Debug.isDebuggerConnected()) {
      val options = CheckOptions(level = "standard", includeDebugInfo = false)
      execute {
        try {
          val result = implementation.performCheck(options)
          val jsResult = Utils.toJSObject(result)
          emitOrBufferSignal(jsResult)
        } catch (e: Exception) {
          Logger.error("Real-time monitor: Debugger check failed: ${e.message}")
        }
      }
    }
  }

  /**
   * Flushes all buffered integrity signals to JavaScript listeners.
   * Ensures FIFO delivery and prevents duplicate emissions.
   *
   * NOTE:
   * - This method is thread-safe and idempotent.
   * - Buffer is cleared immediately after dispatch.
   */
  private fun flushBufferedSignals() {
    synchronized(bufferedSignals) {
      if (hasListeners(EVENT_INTEGRITY_SIGNAL) && bufferedSignals.isNotEmpty()) {
        Logger.debug("Flushing ${bufferedSignals.size} buffered signals to JS")
        val iterator = bufferedSignals.iterator()
        while (iterator.hasNext()) {
          val signal = iterator.next()
          notifyListeners(EVENT_INTEGRITY_SIGNAL, signal, true)
          iterator.remove()
        }
      }
    }
  }

  /**
   * Called when the plugin is being destroyed.
   *
   * CONTRACT:
   * - All BroadcastReceivers registered by this plugin
   *   MUST be unregistered here.
   */
  override fun handleOnDestroy() {
    super.handleOnDestroy()
    unregisterEventReceiver()
  }

  // ---------------------------------------------------------------------------
  // Event emission and buffering
  // ---------------------------------------------------------------------------

  /**
   * Emits an integrity signal to JavaScript or buffers it
   * if no listeners are currently registered.
   *
   * @param signal Fully-formed integrity signal payload.
   */
  private fun emitOrBufferSignal(signal: JSObject) {
    synchronized(bufferedSignals) {
      if (hasListeners(EVENT_INTEGRITY_SIGNAL)) {
        notifyListeners(EVENT_INTEGRITY_SIGNAL, signal, true)
      } else {
        bufferedSignals.add(signal)
      }
    }
  }

  /**
   * Registers the BroadcastReceiver used for passive integrity signals.
   *
   * NOTE:
   * - No polling or background services are introduced.
   * - Observers are scoped strictly to the plugin lifecycle.
   */
  private fun registerEventReceiver() {
    eventReceiver =
      IntegrityEventReceiver { signal ->
        emitOrBufferSignal(signal)
      }

    val intentFilter =
      IntentFilter().apply {
        addAction(Intent.ACTION_PACKAGE_ADDED)
        addAction(Intent.ACTION_PACKAGE_REPLACED)
        addDataScheme("package")
      }

    val flags =
      // UPSIDE_DOWN_CAKE API LEVEL 34
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
        ContextCompat.RECEIVER_NOT_EXPORTED
      } else {
        0
      }

    // TIRAMISU API LEVEL 33
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      context.registerReceiver(eventReceiver, intentFilter, flags)
    } else {
      context.registerReceiver(eventReceiver, intentFilter)
    }
  }

  /**
   * Unregisters the integrity BroadcastReceiver.
   *
   * NOTE:
   * - It is safe to call this method even if the receiver
   *   was never registered.
   */
  private fun unregisterEventReceiver() {
    // It's safe to call unregisterReceiver even if the receiver was never registered
    try {
      context.unregisterReceiver(eventReceiver)
    } catch (_: IllegalArgumentException) {
      // Receiver wasn't registered, ignore
    }
  }

  // ---------------------------------------------------------------------------
  // Error mapping
  // ---------------------------------------------------------------------------

  /**
   * Maps native NativeError values to JavaScript-facing error codes.
   *
   * CONTRACT:
   * - This method is the ONLY place where native errors
   *   are translated into JS-visible failures.
   * - Error codes MUST be:
   *   - stable
   *   - documented
   *   - identical across platforms
   */
  private fun reject(
    call: PluginCall,
    error: NativeError,
  ) {
    val message = error.message ?: ErrorMessages.INTERNAL_ERROR
    call.reject(message, error.errorCode)
  }

  private fun handleError(
    call: PluginCall,
    throwable: Throwable,
  ) {
    if (throwable is NativeError) {
      reject(call, throwable)
    } else {
      val message = throwable.message ?: ErrorMessages.UNEXPECTED_NATIVE_ERROR
      reject(call, NativeError.InitFailed(message))
    }
  }

  // ---------------------------------------------------------------------------
  // Check
  // ---------------------------------------------------------------------------

  /**
   * Executes an integrity check.
   *
   * CONTRACT:
   * - Resolves exactly once on success
   * - Rejects exactly once on failure
   * - Never throws outside this method
   *
   * NOTE:
   * - Option defaulting happens here by design.
   * - The Impl layer MUST receive fully normalized options.
   */
  @PluginMethod
  fun check(call: PluginCall) {
    val options =
      CheckOptions(
        level = call.getString("level") ?: "basic",
        includeDebugInfo = call.getBoolean("includeDebugInfo") ?: false,
      )

    // Execute integrity checks off the plugin thread to avoid future ANR risks.
    execute {
      try {
        val result = implementation.performCheck(options)
        val jsResult = Utils.toJSObject(result)
        call.resolve(jsResult)
      } catch (e: NativeError) {
        handleError(call, e)
      } catch (e: Exception) {
        handleError(call, e)
      }
    }
  }

  // ---------------------------------------------------------------------------
  // PresentBlockPage
  // ---------------------------------------------------------------------------

  /**
   * Presents the configured integrity block page, if enabled.
   *
   * CONTRACT:
   * - This method NEVER decides when it should be called.
   * - The decision is fully delegated to the host application.
   *
   * NOTE:
   * - Returning `{ presented: false }` is NOT an error.
   * - This allows deterministic branching on the JS side.
   *
   * WARNING:
   * - UI navigation is allowed ONLY in the Plugin layer.
   * - The Impl layer MUST NEVER start Activities.
   */
  @PluginMethod
  fun presentBlockPage(call: PluginCall) {
    val blockPage = config.blockPage
    if (blockPage == null || !blockPage.enabled || blockPage.url == null) {
      call.resolve(JSObject().put("presented", false))
      return
    }

    val reason = call.getString("reason")
    val dismissible = call.getBoolean("dismissible") ?: false
    val customUrl = call.getString("customUrl")

    // Apply URL limit validation
    if (customUrl != null && customUrl.length > 2048) {
      handleError(call, NativeError.InvalidInput("customUrl exceeds maximum length of 2048 characters"))
      return
    }

    // Determine final URL: customUrl takes precedence over config
    val urlBase = customUrl ?: blockPage.url

    // Build query string with reason and context
    val queryParams = mutableListOf<String>()

    if (reason != null) {
      queryParams.add("reason=${Uri.encode(reason)}")
    }

    val contextObj = call.getObject("context")
    if (contextObj != null && contextObj.length() > 0) {
      val contextJson = contextObj.toString()
      queryParams.add("context=${Uri.encode(contextJson)}")
    }

    val finalUrl: String
    finalUrl =
      if (queryParams.isEmpty()) {
        urlBase
      } else {
        val queryString = queryParams.joinToString("&")
        if (urlBase.contains("?")) {
          "$urlBase&$queryString"
        } else {
          "$urlBase?$queryString"
        }
      }

    val intent =
      Intent(
        context,
        io.capkit.integrity.ui.BlockActivity::class.java,
      ).apply {
        putExtra("url", finalUrl)
        putExtra("dismissible", dismissible)
        putExtra("preventTapJacking", blockPage.preventTapJacking)
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

        // If not dismissible, clear the back stack
        if (!dismissible) {
          addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
        }
      }

    context.startActivity(intent)

    call.resolve(JSObject().put("presented", true))
  }

  // ---------------------------------------------------------------------------
  // Version
  // ---------------------------------------------------------------------------

  /**
   * Returns the native plugin version.
   *
   * NOTE:
   * - Used exclusively for diagnostics and compatibility checks.
   * - Must not be used for feature detection.
   */
  @PluginMethod
  fun getPluginVersion(call: PluginCall) {
    val ret = JSObject()
    ret.put("version", BuildConfig.PLUGIN_VERSION)
    call.resolve(ret)
  }

  // ---------------------------------------------------------------------------
  // BroadcastReceiver implementation
  // ---------------------------------------------------------------------------

  /**
   * BroadcastReceiver for passive system integrity signals.
   *
   * PURPOSE:
   * - Reacts to system-level events that may indicate integrity changes
   *   (e.g. package installation or replacement).
   *
   * NOTE:
   * - Failures are logged but NOT propagated to JavaScript.
   * - This receiver is observational and non-blocking.
   */
  private inner class IntegrityEventReceiver(
    private val onSignalDetected: (JSObject) -> Unit,
  ) : BroadcastReceiver() {
    override fun onReceive(
      context: Context?,
      intent: Intent?,
    ) {
      when (intent?.action) {
        Intent.ACTION_PACKAGE_ADDED, Intent.ACTION_PACKAGE_REPLACED -> {
          val options =
            CheckOptions(
              level = "standard",
              includeDebugInfo = false,
            )

          // Execute integrity checks off the main/plugin thread to avoid blocking
          execute {
            try {
              val result = implementation.performCheck(options)
              val jsResult = Utils.toJSObject(result)
              onSignalDetected(jsResult)
            } catch (e: NativeError) {
              Logger.error(
                "IntegrityEventReceiver: Error during package change check: ${e.message}",
              )
            } catch (e: Exception) {
              Logger.error(
                "IntegrityEventReceiver: Unexpected error during package change check: ${e.message}",
              )
            }
          }
        }
      }
    }
  }
}

File: packages/integrity/ios/Sources/IntegrityPlugin/IntegrityPlugin.swift
import Foundation
import Capacitor
import UIKit

/**
 Capacitor bridge for the Integrity plugin (iOS).

 Responsibilities:
 - Parse JavaScript input
 - Invoke the native implementation
 - Resolve or reject CAPPluginCall exactly once
 - Map native NativeError to JS-facing error codes

 Forbidden:
 - Platform-specific business logic
 - System API usage
 - Throwing uncaught exceptions
 */
@objc(IntegrityPlugin)
public final class IntegrityPlugin: CAPPlugin, CAPBridgedPlugin {

    // MARK: - Plugin metadata

    /// The unique identifier for the plugin.
    ///
    /// CONTRACT:
    /// - MUST match the plugin registration name used by Capacitor
    /// - MUST remain stable across releases (breaking change otherwise)
    public let identifier = "IntegrityPlugin"

    /// The name used to reference this plugin in JavaScript.
    ///
    /// CONTRACT:
    /// - MUST match `registerPlugin({ name })` on the JS side
    /// - Changing this value breaks JS ↔ native binding
    public let jsName = "Integrity"

    /**
     A list of methods exposed by this plugin.
     - `check`:
     - `presentBlockPage`:
     - `getPluginVersion`: A method that returns the version of the plugin.

     CONTRACT:
     - Every method listed here MUST:
     - be implemented exactly once
     - resolve or reject its CAPPluginCall exactly once
     - return a Promise on the JS side

     NOTE:
     - No method in this list may perform platform logic directly.
     - All business logic MUST be delegated to the Impl layer.
     */
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "check", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "presentBlockPage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    // MARK: - Properties

    /// Native implementation instance.
    private let implementation = Integrity()

    // Configuration instance
    private var config: Config?

    // MARK: - Event-related properties

    /// Buffer for integrity signals captured before a JS listener is registered.
    private var bufferedSignals: [[String: Any]] = []

    /// Canonical event name emitted to the JavaScript layer.
    private static let integritySignalEvent = "integritySignal"

    // MARK: - Lifecycle

    /**
     Plugin lifecycle entry point.

     CONTRACT:
     - Called exactly once by Capacitor
     - This is the ONLY valid place to:
     - read plugin configuration
     - create Config
     - inject configuration into the Impl layer
     - register system event observers (NotificationCenter)

     WARNING:
     - Calling applyConfig(_:) outside this method is FORBIDDEN
     - Observers registered here MUST be detached in `deinit`
     */
    override public func load() {
        // Initialize Config
        let cfg = Config(plugin: self)
        self.config = cfg

        // Log plugin loaded first (before config) - use info since verbose may not be set yet
        Logger.info("Plugin loaded. Version: ", PluginVersion.number)

        implementation.applyConfig(cfg)

        // Register passive system observers
        addEventObservers()
    }

    /**
     Overridden to catch the moment JavaScript starts listening.
     Ensures boot-time signals are delivered immediately.
     */
    override public func addEventListener(_ eventName: String, listener: CAPPluginCall) {
        // Fix: Added explicit 'listener:' label required by the CAPPlugin base class
        super.addEventListener(eventName, listener: listener)

        // Ensure we match the canonical event name used in the plugin
        if eventName == "onIntegritySignal" {
            // Trigger a flush of early boot signals
            let options = CheckOptions(level: "standard", includeDebugInfo: false)
            do {
                // This will internally call mergeBootSignals to clear the volatile queue
                let report = try implementation.performCheck(options: options)
                if let signals = report["signals"] as? [[String: Any]], !signals.isEmpty {
                    emitOrBufferSignal(report)
                }
            } catch {
                Logger.error("Failed to flush boot signals on listener registration")
            }
        }
    }

    /// Plugin teardown.
    ///
    /// NOTE:
    /// - Invoked when the plugin instance is deallocated
    /// - Responsible for detaching all NotificationCenter observers
    deinit {
        removeEventObservers()
    }

    // MARK: - Event emission and buffering

    /**
     Emits an integrity signal to JavaScript or buffers it if no listeners exist.
     */
    private func emitOrBufferSignal(_ signal: [String: Any]) {
        if hasListeners(IntegrityPlugin.integritySignalEvent) {
            notifyListeners(IntegrityPlugin.integritySignalEvent, data: signal, retainUntilConsumed: true)
        } else {
            bufferedSignals.append(signal)
        }
    }

    /**
     Flushes all buffered integrity signals to JavaScript listeners.
     */
    @objc private func flushBufferedSignals() {
        if hasListeners(IntegrityPlugin.integritySignalEvent) && !bufferedSignals.isEmpty {
            for signal in bufferedSignals {
                notifyListeners(IntegrityPlugin.integritySignalEvent, data: signal, retainUntilConsumed: true)
            }
            bufferedSignals.removeAll()
        }
    }

    // MARK: - NotificationCenter registration

    /**
     Registers passive system observers required for real-time integrity signals.
     */
    private func addEventObservers() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleDidBecomeActiveNotification(_:)),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(flushBufferedSignals),
            name: UIApplication.willEnterForegroundNotification,
            object: nil
        )
    }

    /**
     Detaches all NotificationCenter observers.

     NOTE:
     - This method is invoked exclusively from `deinit`.
     - The SwiftLint rule is suppressed intentionally as the call is lifecycle-safe.
     */
    private func removeEventObservers() {
        // swiftlint:disable notification_center_detachment
        NotificationCenter.default.removeObserver(self)
        // swiftlint:enable notification_center_detachment
    }

    // MARK: - Error mapping

    /**
     Maps native `NativeError` values to JS-facing error codes.

     CONTRACT:
     - Error codes MUST be stable and documented
     - Error codes MUST match across platforms
     - Platform-specific error codes are FORBIDDEN
     */
    private func reject(
        _ call: CAPPluginCall,
        error: NativeError
    ) {
        call.reject(error.message, error.errorCode)
    }

    private func handleError(_ call: CAPPluginCall, _ error: Error) {
        if let nativeError = error as? NativeError {
            reject(call, error: nativeError)
        } else {
            let message = error.localizedDescription.isEmpty
                ? ErrorMessages.unexpectedNativeError
                : error.localizedDescription
            reject(call, error: .initFailed(message))
        }
    }

    // MARK: - Integrity check

    /**
     Executes a baseline integrity check.
     */
    @objc func check(_ call: CAPPluginCall) {
        do {
            let options =
                try call.decode(CheckOptions.self)

            let normalizedOptions = CheckOptions(
                level: options.level ?? "basic",
                includeDebugInfo: options.includeDebugInfo ?? false
            )

            let result =
                try implementation.performCheck(
                    options: normalizedOptions
                )

            call.resolve(result)

        } catch let error as NativeError {
            reject(call, error: error)
        } catch {
            handleError(call, error)
        }
    }

    // MARK: - Present block page

    /**
     Presents the configured integrity block page, if enabled.
     */
    @objc func presentBlockPage(_ call: CAPPluginCall) {
        guard
            let blockPage = config?.blockPage,
            blockPage.enabled,
            let baseURL = blockPage.url
        else {
            call.resolve(["presented": false])
            return
        }

        let reason = call.getString("reason")
        let dismissible = call.getBool("dismissible") ?? false
        let customURL = call.getString("customUrl")

        // Apply URL limit validation
        if let custom = customURL, custom.count > 2048 {
            reject(call, error: .invalidInput("customUrl exceeds maximum length of 2048 characters"))
            return
        }

        // Determine final URL: customUrl takes precedence over config
        let urlBase = customURL ?? baseURL

        // Build query string with reason and context
        var queryParams: [String] = []

        if let reason = reason {
            if let encoded = reason.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
                queryParams.append("reason=\(encoded)")
            }
        }

        if let contextObj = call.getObject("context"), !contextObj.isEmpty {
            if let contextData = try? JSONSerialization.data(withJSONObject: contextObj),
               let contextString = String(data: contextData, encoding: .utf8),
               let encoded = contextString.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
                queryParams.append("context=\(encoded)")
            }
        }

        let finalURL: String
        if queryParams.isEmpty {
            finalURL = urlBase
        } else {
            let queryString = queryParams.joined(separator: "&")
            finalURL = urlBase.contains("?")
                ? "\(urlBase)&\(queryString)"
                : "\(urlBase)?\(queryString)"
        }

        DispatchQueue.main.async {
            // CONTRACT:
            // All UIKit interactions MUST occur on the main thread.
            guard let rootVC = self.bridge?.viewController else {
                self.reject(call, error: .unavailable(ErrorMessages.viewControllerUnavailable))
                return
            }

            let blockVC = BlockViewController(
                url: finalURL,
                dismissible: dismissible
            )
            let nav = UINavigationController(rootViewController: blockVC)
            rootVC.present(nav, animated: true) {
                call.resolve(["presented": true])
            }
        }
    }

    // MARK: - Version

    /**
     Returns the native plugin version.

     - Used exclusively for diagnostics and compatibility checks.
     - Must not be used for feature detection.
     */
    @objc func getPluginVersion(_ call: CAPPluginCall) {
        // Standardized enum name across all CapKit plugins
        call.resolve([
            "version": PluginVersion.number
        ])
    }

    // MARK: - Notification handlers

    /**
     Handles application foreground transitions.
     */
    @objc private func handleDidBecomeActiveNotification(_ notification: Notification) {
        // Targeted options for real-time monitoring.
        // Using explicit labels to satisfy Swift compiler and struct definition.
        let options = CheckOptions(
            level: "standard",
            includeDebugInfo: false
        )

        // Perform check and notify JS immediately if signals are found
        do {
            let report = try implementation.performCheck(options: options)

            // Optimization: Only emit if the report contains signals (compromised or score > 0)
            if let signals = report["signals"] as? [[String: Any]], !signals.isEmpty {
                emitOrBufferSignal(report)

                // Logging for internal audit
                Logger.debug("Real-time signal detected and emitted to JS")
            }
        } catch {
            Logger.error("Real-time monitor failed during app activation:", error.localizedDescription)
        }
    }
}
